connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java

```diff
@@ -0,0 +1,277 @@
+        super(new URL[0], parent);
+        this.pluginPaths = pluginPaths;
+        this.pluginLoaders = new ConcurrentHashMap<>();
+        this.activePaths = new HashMap<>();
+        this.inactivePaths = new HashMap<>();
+        this.connectors = new TreeSet<>();
+        this.converters = new TreeSet<>();
+        this.transformations = new TreeSet<>();
+    }
+
+    public DelegatingClassLoader(List<String> pluginPaths) {
+        this(pluginPaths, ClassLoader.getSystemClassLoader());
+    }
+
+    private static boolean isConcrete(Class<?> cls) {
+        final int mod = cls.getModifiers();
+        return !Modifier.isAbstract(mod) && !Modifier.isInterface(mod);
+    }
+
+    private static List<Path> pluginDirs(Path topDir) throws IOException {
+        DirectoryStream.Filter<Path> dirFilter = new DirectoryStream.Filter<Path>() {
+            public boolean accept(Path path) throws IOException {
+                return Files.isDirectory(path);
+            }
+        };
+
+        List<Path> dirs = new ArrayList<>();
+        // Non-recursive for now
+        try (DirectoryStream<Path> listing = Files.newDirectoryStream(topDir, dirFilter)) {
+            for (Path dir : listing) {
+                dirs.add(dir);
+            }
+        }
+        return dirs;
+    }
+
+    private static List<URL> jarPaths(Path pluginDir) throws IOException {
+        DirectoryStream.Filter<Path> jarFilter = new DirectoryStream.Filter<Path>() {
+            public boolean accept(Path file) throws IOException {
+                return file.toString().toLowerCase(Locale.ROOT).endsWith(".jar");
+            }
+        };
+
+        List<URL> jars = new ArrayList<>();
+        try (DirectoryStream<Path> listing = Files.newDirectoryStream(pluginDir, jarFilter)) {
+            for (Path jar : listing) {
+                jars.add(jar.toUri().toURL());
+            }
+        }
+        return jars;
+    }
+
+    public void addAlias(PluginDesc<?> plugin, String alias) {
+        SortedMap<PluginDesc<?>, PluginClassLoader> inner = pluginLoaders.get(plugin.className());
+        if (inner != null) {
+            pluginLoaders.putIfAbsent(alias, inner);
+        }
+    }
+
+    private <T> void addPlugins(Collection<PluginDesc<T>> plugins, PluginClassLoader loader) {
+        for (PluginDesc<T> plugin : plugins) {
+            String pluginClassName = plugin.className();
+            SortedMap<PluginDesc<?>, PluginClassLoader> inner = pluginLoaders.get(pluginClassName);
+            if (inner == null) {
+                inner = new TreeMap<>();
+                pluginLoaders.put(pluginClassName, inner);
+            }
+            inner.put(plugin, loader);
+        }
+    }
+
+    public void initLoaders() {
+        for (String path : pluginPaths) {
+            try {
+                Path pluginPath = Paths.get(path).toAbsolutePath();
+                if (Files.isDirectory(pluginPath)) {
```

Comment:
Yes, as I mentioned above, the initial draft support only dirs of jars, but we could easily add uberjars and directories of classes. Will look into it.